app/gui2/src/components/GraphEditor/widgets/WidgetFileBrowser.vue

```diff
@@ -0,0 +1,119 @@
+<script setup lang="ts">
+import NodeWidget from '@/components/GraphEditor/NodeWidget.vue'
+import {
+  CustomDropdownItemsKey,
+  type CustomDropdownItem,
+} from '@/components/GraphEditor/widgets/WidgetSelection.vue'
+import { Score, WidgetInput, defineWidget, widgetProps } from '@/providers/widgetRegistry'
+import { useGraphStore } from '@/stores/graph'
+import type { RequiredImport } from '@/stores/graph/imports'
+import { Ast } from '@/util/ast'
+import { ArgumentInfoKey } from '@/util/callTree'
+import { computed } from 'vue'
+
+const props = defineProps(widgetProps(widgetDefinition))
+const graph = useGraphStore()
+
+const insertAsFileConstructor = computed(() => {
+  const reprType = props.input[ArgumentInfoKey]?.info?.reprType
+  if (reprType == null) return false
+  const textTypePos = reprType.indexOf(TEXT_TYPE)
+  const fileTypePos = reprType.indexOf(FILE_TYPE)
+  // If both types are present, check if TEXT_TYPE comes after FILE_TYPE.
+  if (fileTypePos !== -1 && (textTypePos === -1 || textTypePos > fileTypePos)) {
+    return true
+  } else {
+    return false
+  }
+})
+const strictlyFile = computed(() => props.input.dynamicConfig?.kind === 'File_Browse')
+const strictlyDirectory = computed(() => props.input.dynamicConfig?.kind === 'Folder_Browse')
+const label = computed(() => (strictlyDirectory.value ? 'Choose directory…' : 'Choose file…'))
+
+const FILE_CONSTRUCTOR = FILE_TYPE + '.new'
+const FILE_SHORT_CONSTRUCTOR = 'File.new'
+
+function makeValue(edit: Ast.MutableModule, useFileConstructor: boolean, path: string): Ast.Owned {
+  if (useFileConstructor) {
+    const arg = Ast.TextLiteral.new(path, edit)
+    const requiredImport = {
+      kind: 'Unqualified',
+      from: FILE_MODULE,
+      import: 'File',
+    } as RequiredImport
+    const conflicts = graph.addMissingImports(edit, [requiredImport])
+    const constructor = conflicts != null ? FILE_CONSTRUCTOR : FILE_SHORT_CONSTRUCTOR
+    const constructorAst = Ast.PropertyAccess.tryParse(constructor, edit)
+    if (constructorAst == null) {
+      throw new Error(`Failed to parse constructor as AST: ${constructor}`)
+    }
+    return Ast.App.new(edit, constructorAst, undefined, arg)
+  } else {
+    return Ast.TextLiteral.new(path, edit)
+  }
+}
+
+const item = computed<CustomDropdownItem>(() => ({
```

Comment:
The only reactive part here is the label, right? I think it would be more clear to define the `onClick` function outside the `computed` so that it's obvious it isn't capturing anything reactively.

Reply:
Thanks, fixed.